### PR TITLE
Many to many relationship

### PR DESCRIPTION
**Pull request by alfonsserra**
Add a many to many relationship.
In this case the involved entities are Allergy, and Allergies for a patient.
Use new application frame UI

### CODE REVIEW

File: src/common/components/allergy/combobox/allergy-combobox.component.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+	templateUrl: '../../../../../node_modules/systelab-components/html/abstract-combobox.component.html'
+})
+
+export class AllergyComboBox extends AbstractApiComboBox<Allergy> {
+
+	private totalItems = 0;
+	private _startsWith: string;
+
+	@Input()
+	set startsWith(value: string) {
```
Comment (carlosra85) [inline src/common/components/allergy/combobox/allergy-combobox.component.ts]: startsWith not used

File: src/features/patient-maintenance/patient-allergy-detail-dialog/patient-allergy-dialog.component.html [hunk 2]
```diff
@@ -0,0 +1,38 @@
+<systelab-dialog-header (close)="close()">{{title}}</systelab-dialog-header>
+<div class="slab-flex-1">
+
+    <form class="p-2">
+        <div class="row mt-1">
+            <label class="col-sm-3 col-form-label text-truncate">{{'COMMON_ALLERGY' | translate | async}}</label>
+            <div class="col-sm-9">
+                <allergy-combobox [(id)]="allergyId" [(description)]="allergyName" [isDisabled]="isUpdate()"></allergy-combobox>
```
Comment (carlosra85) [inline src/features/patient-maintenance/patient-allergy-detail-dialog/patient-allergy-dialog.component.html]: Normally description does not need to emit

File: src/common/components/patient-allergy/grid/patient-allergy-grid.component.ts [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+	private totalItems = 0;
+
+	constructor(protected preferencesService: PreferencesService, protected i18nService: I18nService, protected dialogService: DialogService, protected patientAllergyService: PatientAllergyService) {
+		super(preferencesService, i18nService, dialogService);
+	}
+
+	protected getColumnDefs(): Array<any> {
+		const columnDefs: Array<any> = [];
+		this.i18nService.get(['COMMON_NAME', 'COMMON_ASSERTED_DATE', 'COMMON_NOTES'])
```
Comment (vinyulis) [inline src/common/components/patient-allergy/grid/patient-allergy-grid.component.ts]: Use i18nService instant inside colDef

File: src/common/components/patient-allergy/grid/patient-allergy-grid.component.ts [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+							return this.getAssertedDate(params.data);
+						}, width: 200
+					},
+					{colId: 'note', headerName: res.COMMON_NOTES, field: 'note', width: 200});
+			});
+		return columnDefs;
+	}
+
+	private getAllergyName(patientAllergy: PatientAllergy): string {
+		return (patientAllergy && patientAllergy.allergy) ? patientAllergy.allergy.name : '';
```
Comment (vinyulis) [inline src/common/components/patient-allergy/grid/patient-allergy-grid.component.ts]: Is needed patientAllergy? Enough with patientAllergy.allergy?
 in

File: src/features/allergy-maintenance/alergy-detail-dialog/allergy-dialog.component.ts [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+	private updateAllergy(allergy: Allergy) {
+		this.allergyService.updateAllergy(allergy.id, allergy)
+			.subscribe((result) => {
+					this.dialog.close(true);
+				},
+				(error) => this.showError(error));
+	}
+
+	private showError(error: HttpErrorResponse) {
```
Comment (vinyulis) [inline src/features/allergy-maintenance/alergy-detail-dialog/allergy-dialog.component.ts]: If we are always showing an error returned by the api maybe is better to have this method in a common or util class

File: src/features/allergy-maintenance/allergy-maintenance.component.ts [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+		}
+	}
+
+	private updateAllergy(contextMenuActionData: GridContextMenuActionData<Allergy>) {
+		this.doSelect(contextMenuActionData.data);
+	}
+
+	private deleteAllergy(contextMenuActionData: GridContextMenuActionData<Allergy>) {
+		this.allergyService.removeAllergy(contextMenuActionData.data.id)
+			.subscribe(result => this.allergygrid.refresh(), error => console.log('error'));
```
Comment (vinyulis) [inline src/features/allergy-maintenance/allergy-maintenance.component.ts]: console.log?

File: src/features/patient-maintenance/allergies-form/patient-allergies-form.component.ts [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+			});
+	}
+
+	private updatePatientAllergy(contextMenuActionData: GridContextMenuActionData<PatientAllergy>) {
+		this.doSelect(contextMenuActionData.data);
+	}
+
+	private deletePatientAllergy(contextMenuActionData: GridContextMenuActionData<PatientAllergy>) {
+		this.patientAllergyService.deleteAllergiesFromPatient(this.patientId, contextMenuActionData.data.allergy.id)
+			.subscribe(result => this.grid.refresh(), error => console.log('error'));
```
Comment (vinyulis) [inline src/features/patient-maintenance/allergies-form/patient-allergies-form.component.ts]: console.log?

File: src/features/patient-maintenance/patient-allergy-detail-dialog/patient-allergy-dialog.component.ts [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+
+	private updatePatientAllergy(patientAllergy: PatientAllergy) {
+		this.patientAllergyService.updateAllergyFromPatient(this.parameters.patientId, patientAllergy.allergy.id, patientAllergy)
+			.subscribe((result) => {
+					this.dialog.close(true);
+				},
+				(error) => this.showError(error));
+	}
+
+	private showError(error: HttpErrorResponse) {
```
Comment (vinyulis) [inline src/features/patient-maintenance/patient-allergy-detail-dialog/patient-allergy-dialog.component.ts]: Same a previous

File: src/app/main/main.component.html [hunk 9]
```diff
@@ -1,11 +1,13 @@
...
-    </p>
-  </div>
+<systelab-app-frame [userName]="userName" [userFullName]="userFullName" [title]="hospitalName" [menu]="menu" [tabs]="sidetabs" [logoIcon]="logoIcon"
+                    [actions]="sideactions" [largeSideBar]="false" (selected)="doSelect($event)">
+    <div header-content class="row justify-content-end mr-3">
+        <button type="button" class="btn float-left" (click)="doPatientSearch();">{{'COMMON_PATIENT' | translate | async}}</button>
+    </div>
+    <div main-content class="slab-flex-1 d-flex">
+        <systelab-tabs *ngIf="showConfiguration" class="slab-flex-1 mt-2" [showTabBackground]="false">
+            <systelab-tab [title]="'Allergies'" [id]="'FromTab'">
```
Comment (apenalver2) [inline src/app/main/main.component.html]: Use something like {{'COMMON_ALLERGIES' | translate | async}} instead of a hardcode literal

**Comment by apenalver2** [comment]
I saw that you removed all this.i18nService.get for the instant, remember if the screen refreshes the instant is not working and considering that this is a seed to play and to learn, in my opinion, people need to have both examples in the code.

Apart from that, is OK, thank you

**Comment by alfonsserra** [comment]
According to @apenalver2 comments, I have used this.i18nService.get instead of this.i18nService.instant

**Pull request merged**